src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_RefReadOnly.cs

```diff
@@ -570,25 +986,734 @@ public class ReadOnlyAttribute : System.Attribute { }
+        }
+
+        [Fact]
+        public void TryingPossibleBindingsForRefReadOnlyDoesNotPolluteCompilationForInvalidOnes()
+        {
+            var reference = CreateCompilationWithMscorlib(@"
+public delegate ref readonly int D1 ();
+public delegate ref int D2 ();
+").VerifyEmitDiagnostics();
+
+            Assert.True(reference.NeedsGeneratedIsReadOnlyAttribute);
+
+            var compilation = CreateCompilationWithMscorlibAndSystemCore(@"
+public class Test
+{
+    public void Process(D1 lambda, int x) { }
+    public void Process(D2 lambda, byte x) { }
+
+    void User()
+    {
+        byte byteVar = 0;
+        Process(() => { throw null; }, byteVar);
+    }
+}", references: new[] { reference.ToMetadataReference() });
+
+            compilation.VerifyEmitDiagnostics();
+            Assert.False(compilation.NeedsGeneratedIsReadOnlyAttribute);
+        }
+
+        [Fact]
+        public void RefReadOnlyErrorsForLambdasDoNotPolluteCompilationDeclarationsDiagnostics()
+        {
+            var reference = CreateCompilationWithMscorlib(@"
+public delegate int D (ref readonly int x);
+").EmitToImageReference();
+
+            var code = @"
+public class Test
+{
+    public void Process(D lambda) { }
+
+    void User()
+    {
+        Process((ref readonly int p) => p);
+    }
+}";
+
+            var compilation = CreateCompilationWithMscorlib(code, options: TestOptions.ReleaseModule, references: new[] { reference });
+
+            compilation.DeclarationDiagnostics.Verify();
+
+            compilation.VerifyDiagnostics(
+                // (8,18): error CS0518: Predefined type 'System.Runtime.CompilerServices.IsReadOnlyAttribute' is not defined or imported
+                //         Process((ref readonly int p) => p);
+                Diagnostic(ErrorCode.ERR_PredefinedTypeNotFound, "ref readonly int p").WithArguments("System.Runtime.CompilerServices.IsReadOnlyAttribute").WithLocation(8, 18));
+        }
+
+        [Fact]
+        public void RefReadOnlyErrorsForLocalFunctionsDoNotPolluteCompilationDeclarationsDiagnostics()
+        {
+            var code = @"
+public class Test
+{
+    private int x = 0;
+    void User()
+    {
+        void local(ref readonly int x) { }
+        local(x);
+    }
+}";
+
+            var compilation = CreateCompilationWithMscorlib(code, options: TestOptions.ReleaseModule);
+
+            compilation.DeclarationDiagnostics.Verify();
+
+            compilation.VerifyDiagnostics(
+                // (7,20): error CS0518: Predefined type 'System.Runtime.CompilerServices.IsReadOnlyAttribute' is not defined or imported
+                //         void local(ref readonly int x) { }
+                Diagnostic(ErrorCode.ERR_PredefinedTypeNotFound, "ref readonly int x").WithArguments("System.Runtime.CompilerServices.IsReadOnlyAttribute").WithLocation(7, 20));
+        }
+
+        [Fact(Skip = "https://github.com/dotnet/roslyn/issues/19395")]
```

Comment:
Enabling it (in a way that doesn't fail the build) means that it will assert no diagnostics are produced, which is the wrong behavior?

---
In reply to: [115871705](https://github.com/dotnet/roslyn/pull/18715#discussion_r115871705) [](ancestors = 115871705)